### PR TITLE
Fixing a 8.7 regression of ring_simplify in ArithRing

### PR DESCRIPTION
**Pull request by herbelin**
This is a PR to continue discussing a solution to the 8.7 regression of `ring_simplify` in `ArithRing`.

@silene: I made the experience of using a "proxy" variant of `to_nat` and then to only post-unfold the proxy copy and it seems to work. I'm not fully sure that this kind of approach is very clean and scales very well (what if a user uses the `proxy` itself, ...) but my time being limited, I can propose it as a fix if noone has a better idea (and if the `11` as `Rec` rather than `Eval` is really a bug).

What do you think?

### CODE REVIEW

File: plugins/setoid_ring/ArithRing.v [hunk 1]
```diff
@@ -19,9 +19,11 @@ Lemma natSRth : semi_ring_theory O (S O) plus mult (@eq nat).
   exact mult_plus_distr_r.
  Qed.
 
+Definition proxy_to_nat := N.to_nat.
```
Comment (SkySkimmer) [inline plugins/setoid_ring/ArithRing.v]: Making it `Local Definition` would make it less likely to be used by mistake right?

Comment (herbelin) [reply plugins/setoid_ring/ArithRing.v]: Good idea. Change committed.

**Review by silene — APPROVED** [review]
The approach makes sense to me. Thanks for the patch.

File: plugins/setoid_ring/ArithRing.v [hunk 2]
```diff
@@ -19,9 +19,11 @@ Lemma natSRth : semi_ring_theory O (S O) plus mult (@eq nat).
   exact mult_plus_distr_r.
  Qed.
 
+Local Definition proxy_to_nat := N.to_nat.
+
 Lemma nat_morph_N :
    semi_morph 0 1 plus mult (eq (A:=nat))
-          0%N 1%N N.add N.mul N.eqb N.to_nat.
+          0%N 1%N N.add N.mul N.eqb proxy_to_nat.
```
Comment (JasonGross) [inline plugins/setoid_ring/ArithRing.v]: I think this is the wrong approach, in particular the use of a `Local` definition in the type of a non local lemma.  Why not switch the sense of proxy, and add as a preprocessing step `change N.to_nat with proxy_to_nat in *`?  Then you could compute on any occurrence of `N.to_nat` in the postprocessing phase, after which you unfold `proxy_to_nat`.  This way, users can't accidentally use `proxy_to_nat`, the way they might here by using `nat_morph_N`.

**Review by JasonGross — DISMISSED** [review]

File: plugins/setoid_ring/ArithRing.v (hunk 2 above)
Comment (herbelin) [reply plugins/setoid_ring/ArithRing.v]: Still a (horrible) hack, but more local indeed. Pushed.

Further questions (from someone who doesn't know well `ring`):
- will this scale to other user-level rings defined on the model of `ArithRing`?
- wouldn't it be clearer, say, e.g. that `Add Ring` has an argument which provides with a postprocessing of constants (somehow a converse to the `constants` field)? Then, maybe, only the declaration of the `R` ring could be changed, and a compatibility for other ring would be kept?

**Comment by silene** [comment]
> wouldn't it be clearer, say, e.g. that `Add Ring` has an argument which provides with a postprocessing of constants (somehow a converse to the constants field)? Then, maybe, only the declaration of the `R` ring could be changed, and a compatibility for other ring would be kept?

If you do not provide a morphism when using `Add Ring`, then a default morphism is used (all the `gen_phi` functions), which expresses a constant using just `1`. So one can wonder whether it is actually a good thing from the user point of view. For `R`, it was working fine because the pretty-printer for real literals was actually the converse of the default morphism for rings. Note how lucky this coincidence was; the slightest difference would have caused a mess for users of real numbers. Obviously, luck had no play here; developers forced the default morphism for _every_ ring just so that it would work for a single one: real numbers.

Now, let us consider the morphism used for `nat`, which is `N.to_nat`. Why wasn't the default morphism used? Because we do not have the above coincidence in that case, that is, the default morphism does not match the pretty-printer for `nat`. So a specific morphism was chosen just to match the pretty-printer. Actually, that is not quite right, since the pretty-printer for `nat` expects fully-reduced numbers and `ring_simplify` was forcefully reducing morphisms, it would have worked anyway, whichever morphism was chosen.

But this begs another question: does it actually make sense to reduce `N.to_nat`? We all know how wrong that is to represent a number using just `S` and `O`. In fact, Coq even feels compelled to warn the users how wrong it is: _Stack overflow or segmentation fault happens when working with large numbers in nat_. Yet `ring_simplify` precisely does what we tell the users not to do. It turns compact literals into exponentially large ones.

Short version: more often that not, not reducing the morphisms is actually the sensible solution. The fact that `ring_simplify` systematically reduced morphisms was a mistake, in my opinion. So this patch is fine in that it restores compatibility. But I think that, in the long run, preserving this compatibility is a mistake and we should switch to a compact representation of `nat` literals, as is done in other proof assistants, e.g. HOL Light.

**Comment by herbelin** [comment]
Failures are unrelated:
- VST: The job exceeded the maximum time limit for jobs, and has been terminated.
- test-suite: coq-makefile/timing/run.sh...Error!
- macos x: shell_session_update: command not found

@silene: Thanks for the explanation.

**Comment by herbelin** [comment]
Hi Jason, I'm dismissing (i.e. "no-longer-valid"-ing) your review since otherwise it seems that @maximedenes' filter don't see it.
